executor/src/witgen/machines/multiplicity_counter.rs

```diff
@@ -0,0 +1,73 @@
+use std::collections::BTreeMap;
+
+use powdr_ast::analyzed::PolyID;
+use powdr_number::{DegreeType, FieldElement};
+
+/// Utility for counting the multiplicities for lookups.
+pub struct MultiplicityCounter {
+    /// Maps an identity ID to the corresponding multiplicity column.
+    /// Note that multiple identity IDs can map to the same multiplicity column.
+    identity_id_to_multiplicity_column: BTreeMap<u64, PolyID>,
+
+    /// A map poly_id -> (index -> count).
+    /// Has a (possibly entry) entry for each value in `identity_id_to_multiplicity_column`.
+    counts: BTreeMap<PolyID, BTreeMap<usize, usize>>,
+}
+
+impl MultiplicityCounter {
+    pub fn new(identity_id_to_multiplicity_column: BTreeMap<u64, PolyID>) -> Self {
+        let counts = identity_id_to_multiplicity_column
+            .values()
+            .cloned()
+            .map(|poly_id| (poly_id, BTreeMap::new()))
+            .collect();
+        Self {
+            identity_id_to_multiplicity_column,
+            counts,
+        }
+    }
+
+    /// For a given identity ID, increments the count of the corresponding multiplicity column at the given index.
+    pub fn increment(&mut self, identity_id: u64, index: usize) {
+        if let Some(poly_id) = self.identity_id_to_multiplicity_column.get(&identity_id) {
```

Comment:
Shouldn't this panic if the identity is not in there instead of currently silently continuing?